### PR TITLE
Auto categories

### PR DESCRIPTION
**Pull request by zero-24**
Hi @mbabker @b2z @elkuku 

Here is just one more new `feature` that would very helpful to me ;)

This change should add more automagic this time to the categories. The Bot should add the "category" based on some tests (changed files) for pull requests as it is easyer to extract it from the pull than from the text ;)

As I have no online system to test availiable (simple no root server in the internet currently) can you review and test, suggest, improve here? Maybe we can use again @b2z 's test site (if he is back from vacation)?

#### Two things I'm not sure

What is the differenz between

`$category['issue_id']` and `$category['issue_number']`?

```php
$categoryModel            = new CategoryModel($this->getContainer()->get('db'));
$category['issue_id']     = $issueNumber;
$category['modified_by']  = $this->getGithubBotName($project);
$category['categories']   = $addCategories;
$category['issue_number'] = $issueNumber;
$category['project_id']   = $project->project_id;

$categoryModel->updateCategory($category);
```

As we use the CategoryModel that handle the database changes we don't need to pass the `IssuesTable $table`every time? Correct?

Thanks for any feedback :smile:

### CODE REVIEW

**Comment by zero-24** [comment]
If you have more ideas for folders / files to check let me know ;)

For a unknwon reason it fails on the changes in the `AbstractListener.php`? Did i do something wrong there?

**Comment by b2z** [comment]
I am almost back :) Need some time to get things done in real life.

> What is the differenz between
> $category['issue_id'] and $category['issue_number']?

`issue_id` is the id of an issue in `#__issues` table and it is not equal to `issue_number`. So you can not assign `$issueNumber` to `$category['issue_id']`.

**Comment by zero-24** [comment]
Ok thanks @b2z So we can use here `$table->id` ;) And the seccond question is also fixed. Thanks :smile:

**Comment by b2z** [comment]
> Ok thanks @b2z So we can use here $table->id ;)
Yeap as it is being passed.

**Comment by b2z** [comment]
Ok [here](http://pear.php.net/bugs/bug.php?id=18815) I found regarding failed PHPCS check - there is a syntax error and PHPCS is failing. That's because you have `protected getGithubBotName` - missing word function :wink:

File: src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php [hunk 1]
```diff
@@ -289,4 +287,105 @@ protected function createStatus(Github $gitHub, TrackerProject $project, $issueN
...
+					'Error retrieving issue number for %s/%s',
+					$project->gh_user,
+					$project->gh_project
+				)
+			);
+
+			throw new \RuntimeException('Error retrieving issue number for ' . $project->gh_user . '/' . $project->gh_project);
+		}
+
+		$categoryModel            = new CategoryModel($this->getContainer()->get('db'));
```
Comment (b2z) [inline src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php]: You are using `IssuesTable` and `CategoryModel` so need to add:
```
use App\Tracker\Model\CategoryModel;
use App\Tracker\Table\IssuesTable;
```

Before `use App\Projects\TrackerProject;`

File: src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php [hunk 2]
```diff
@@ -289,4 +287,105 @@ protected function createStatus(Github $gitHub, TrackerProject $project, $issueN
...
+
+	/**
+	 * Get the files modified by the pull request
+	 *
+	 * @param   object  $hookData   Hook data payload
+	 * @param   Github  $github     Github object
+	 * @param   Logger  $logger     Logger object
+	 * @param   object  $project    Object containing project data
+	 *
+	 * @return  void
```
Comment (b2z) [inline src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php]: Method returns array.

File: src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php [hunk 3]
```diff
@@ -289,4 +287,105 @@ protected function createStatus(Github $gitHub, TrackerProject $project, $issueN
...
+
+	/**
+	 * Return the currently configured bot account. Fallback is 'joomla-cms-bot'
+	 *
+	 * @return  string  The currently configured bot account
+	 *
+	 * @since   1.0
+	 *
+	 */
+	protected getGithubBotName($project)
```
Comment (b2z) [inline src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php]: Missing word `function` here.

File: src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php [hunk 4]
```diff
@@ -289,4 +287,105 @@ protected function createStatus(Github $gitHub, TrackerProject $project, $issueN
...
+			$files = array();
+		}
+
+		// Retrun the changed files
+		return $files;
+	}
+
+	/**
+	 * Return the currently configured bot account. Fallback is 'joomla-cms-bot'
+	 *
```
Comment (b2z) [inline src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php]: Missing doc comment for `$project`

**Comment by zero-24** [comment]
Just fixed the first issues and travis again. thanks @b2z

File: src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php (hunk 1 above)
Comment (b2z) [inline src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php]: Well another problem here - there is no `getContainer()` method in `AbstractListener` class :(

Comment (zero-24) [inline src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php]: hmm correct but i can't see it here too:
https://github.com/joomla/jissues/blob/master/src/App/Tracker/Controller/Issue/Save.php#L199-L206

Comment (zero-24) [inline src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php]: I'm confused?

Comment (b2z) [inline src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php]: Yes you are :wink: The class you referenced extends `AbstractTrackerController` which is container aware. [See these lines](https://github.com/joomla/jissues/blob/master/src/JTracker/Controller/AbstractTrackerController.php#L29-L31). May be we can make our AbstractListener also to be container aware

File: src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php [hunk 5]
```diff
@@ -289,4 +286,107 @@ protected function createStatus(Github $gitHub, TrackerProject $project, $issueN
...
+		}
+
+		$categoryModel            = new CategoryModel($this->getContainer()->get('db'));
+		$category['issue_id']     = $table->id;
+		$category['modified_by']  = $this->getGithubBotName($project);
+		$category['categories']   = $addCategories;
+		$category['issue_number'] = $issueNumber;
+		$category['project_id']   = $project->project_id;
+
+		$categoryModel->updateCategory($category);
```
Comment (b2z) [inline src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php]: Looking on the `updateCategory()` method of `CategoryModel` I see that it accepts array with category Ids, but we are passing category aliases that were set in `checkFilesAndAssignCategory()`.

File: src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php (hunk 1 above)
Comment (zero-24) [inline src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php]: So we need this lines:

````
use Joomla\DI\ContainerAwareInterface;
use Joomla\DI\ContainerAwareTrait;
```

Plus

```
use ContainerAwareTrait;
```

And us this:
`abstract class AbstractListener implements ContainerAwareInterface`

?

Comment (b2z) [inline src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php]: Yeap :)

File: src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php (hunk 5 above)
Comment (zero-24) [inline src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php]: hmm i can't see the IDs in the jissues backed can you send me a list of category to id? (e.g. via Glip / mail?

File: src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php (hunk 1 above)
Comment (zero-24) [inline src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php]: Done :smile: 

https://github.com/zero-24/jissues/commit/c1f10efee8e9530e0cff572d8b98545ffdcd4019

File: src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php (hunk 5 above)
Comment (b2z) [inline src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php]: Do you have access to Categories manager there? You can see the id in the edit link.

Comment (zero-24) [inline src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php]: Yes i have. ok got it ;) Give me a few minutes to fix it up ;)

Comment (zero-24) [inline src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php]: just done see: https://github.com/joomla/jissues/commit/3cbe0177632626e1d9efbf79071e670598d83f5e

Comment (b2z) [inline src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php]: Travis is not happy again :)

File: src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php (hunk 1 above)
Comment (mbabker) [inline src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php]: To make it container aware you also have to inject the container in when the classes get instantiated.  The ContainerAwareInterface just defines some required methods to interact with the container within an object.

File: src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php (hunk 5 above)
Comment (zero-24) [inline src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php]: >Travis should wirte my PRs than he is allways happy with it .. :)

Hopfully he is happy again ;) https://github.com/joomla/jissues/commit/db1d2fe6f30d4bc2c9d35c883b4fd2311f14d6b4

File: src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php (hunk 1 above)
Comment (zero-24) [inline src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php]: how can this be done?

Do you referer to something like: https://github.com/joomla/jissues/blob/master/src/JTracker/Controller/AbstractTrackerController.php#L97-L119?

Comment (mbabker) [inline src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php]: Basically in the [addEventListener](https://github.com/joomla/jissues/blob/master/src/JTracker/Controller/AbstractTrackerController.php#L462) method you'll need to change the inside of the `if class_exists()` check to this:

```php
$listener = new $fullClass;
$listener->setContainer($this->getContainer());
$this->dispatcher->addListener($listener);
$this->listenerSet = true;
```

Comment (zero-24) [inline src/App/Tracker/Controller/Hooks/Listeners/AbstractListener.php]: Done here: https://github.com/joomla/jissues/commit/a218bf8a20b8064ca13990bbd9d679e4af7813c9

**Comment by b2z** [comment]
So how we will test it? Ids of CMS categories are not the same as categories on my test server :)

**Comment by zero-24** [comment]
Lets add a category with id `2` and create this directory `administrator/components/com_admin/sql/updates/postgresql`

https://github.com/joomla/jissues/pull/705/files#diff-e6200f2e3305ca89413408a6216286efR763

If that work it looks very well. Than we can try also id 3 with `administrator/components/com_admin/sql/updates/sqlazure`

and than we can try to create or edit a file in both directory. it that work i expect that it work ;)

**Comment by b2z** [comment]
Ok, but first we should test #696 - I've switched test server to it.

**Comment by b2z** [comment]
Ok I've switched test server to this PR and set up needed categories in http://jissues.j4devs.ru/tracker/jissues-test. Now we can try to issue PRs against https://github.com/b2z/jissues-test ;)

**Comment by zero-24** [comment]
@b2z can you confirm that you rename the listener for your envoirment? As it currently don't work see: http://jissues.j4devs.ru/tracker/jissues-test/23

**Comment by b2z** [comment]
Yeah forgot to update pulls listener. Now should be fine.

**Comment by zero-24** [comment]
@b2z can you try the last commit? I have implemented the `in_array` check wrong :(

**Comment by b2z** [comment]
Test server is updated

**Comment by zero-24** [comment]
hmm can you see any error logged? I have no idea why it fails :(

**Comment by b2z** [comment]
I have only 
```
[Wed Oct 14 17:04:11 2015] [error] [client 192.30.252.40] PHP Notice:  Undefined property: stdClass::$issue in /home/admin/web/jissues.j4devs.ru/jissues/src/App/Track
er/Controller/Hooks/ReceiveIssuesHook.php on line 42
[Wed Oct 14 17:04:11 2015] [error] [client 192.30.252.40] PHP Notice:  Trying to get property of non-object in /home/admin/web/jissues.j4devs.ru/jissues/src/App/Track
er/Controller/Hooks/ReceiveIssuesHook.php on line 42
[Wed Oct 14 17:04:11 2015] [error] [client 192.30.252.40] PHP Notice:  Undefined property: stdClass::$action in /home/admin/web/jissues.j4devs.ru/jissues/src/App/Trac
ker/Controller/Hooks/ReceiveIssuesHook.php on line 64
[Wed Oct 14 17:04:11 2015] [error] [client 192.30.252.40] PHP Notice:  Undefined property: stdClass::$issue in /home/admin/web/jissues.j4devs.ru/jissues/src/App/Track
er/Controller/Hooks/ReceiveIssuesHook.php on line 68
```

But not fatal errors or something :(

**Comment by b2z** [comment]
It should be debugged somehow, seems that this will be a job for local environment.

**Pull request closed without merging**